### PR TITLE
[WIP] added deploy key resource

### PR DESCRIPTION
**Pull request by sjauld**
My first crack at adding a new resource - please let me know if anything else is needed.

```
make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?   	github.com/terraform-providers/terraform-provider-bitbucket	[no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccBitbucketDefaultReviewers_basic
--- PASS: TestAccBitbucketDefaultReviewers_basic (19.40s)
=== RUN   TestAccBitbucketDeployKey_basic
--- PASS: TestAccBitbucketDeployKey_basic (20.88s)
=== RUN   TestAccBitbucketHook_basic
--- PASS: TestAccBitbucketHook_basic (22.34s)
=== RUN   TestAccBitbucketRepository_basic
--- PASS: TestAccBitbucketRepository_basic (10.86s)
=== RUN   TestAccBitbucketRepository_camelcase
--- PASS: TestAccBitbucketRepository_camelcase (11.90s)
PASS
ok  	github.com/terraform-providers/terraform-provider-bitbucket/bitbucket	85.394s
```

### CODE REVIEW

File: bitbucket/resource_deploy_key.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+func resourceDeployKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDeployKeyCreate,
+		Read:   resourceDeployKeyRead,
+		Update: resourceDeployKeyUpdate,
+		Delete: resourceDeployKeyDelete,
+		Exists: resourceDeployKeyExists,
+
+		Schema: map[string]*schema.Schema{
+			"owner": &schema.Schema{
```
Comment (pdecat) [inline bitbucket/resource_deploy_key.go:31]: The API documentation refers to this as `accountname`which is what I used in my own PR  https://github.com/terraform-providers/terraform-provider-bitbucket/pull/6/files#diff-212cd6fa62f60904d8a923069edaa6c9R43

Not sure which is better.

File: bitbucket/provider_test.go [hunk 2]
```diff
@@ -31,9 +31,12 @@ func TestProvider_impl(t *testing.T) {
...
 func testAccPreCheck(t *testing.T) {
 	if v := os.Getenv("BITBUCKET_USERNAME"); v == "" {
-		t.Fatal("BITBUCKET_USERNAME must be set for acceptence tests")
+		t.Fatal("BITBUCKET_USERNAME must be set for acceptance tests")
 	}
 	if v := os.Getenv("BITBUCKET_PASSWORD"); v == "" {
-		t.Fatal("BITBUCKET_PASSWORD must be set for acceptence tests")
+		t.Fatal("BITBUCKET_PASSWORD must be set for acceptance tests")
+	}
+	if v := os.Getenv("BITBUCKET_TEAM"); v == "" {
```
Comment (pdecat) [inline bitbucket/provider_test.go:39]: Not that my opinion matters at all, but I like the idea of introducing a `BITBUCKET_TEAM` environment variable.

File: bitbucket/resource_deploy_key.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	return resourceDeployKeyRead(d, m)
+}
+
+func resourceDeployKeyRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(*BitbucketClient)
+
+	deployKey_req, _ := client.Get(fmt.Sprintf("1.0/repositories/%s/%s/deploy-keys/%s",
+		d.Get("owner").(string),
+		d.Get("repository").(string),
+		url.PathEscape(d.Id()),
```
Comment (pdecat) [inline bitbucket/resource_deploy_key.go]: Why escape as the pk/id is an integer?

File: bitbucket/resource_deploy_key.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	))
+
+	log.Printf("ID: %s", url.PathEscape(d.Id()))
```
Comment (pdecat) [inline bitbucket/resource_deploy_key.go]: Shouldn't that be in debug level?

File: bitbucket/resource_deploy_key.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+	payload, err := json.Marshal(deployKey)
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Put(fmt.Sprintf("1.0/repositories/%s/%s/deploy-keys/%s",
+		d.Get("owner").(string),
+		d.Get("repository").(string),
+		url.PathEscape(d.Id()),
```
Comment (pdecat) [inline bitbucket/resource_deploy_key.go]: Why escape as the pk/id is an integer?

**Review by pdecat — COMMENTED** [review]
I'm not a maintainer but as I also submitted my first PR to add a group resource using the 1.0 API a few weeks ago, I was curious to see if we did things differently or not.

Anyway, here is my informal review.

File: bitbucket/resource_deploy_key.go (hunk 1 above)
Comment (sjauld) [reply bitbucket/resource_deploy_key.go:31]: Yeah I used `owner` to align with the existing resources in this repo. Atlassian's docs seem to interchangeably use `accountname` or `username` (maybe a 1.0 vs 2.0 difference), so I guess `owner` is as good as anything.

**Comment by sjauld** [comment]
⏳

File: bitbucket/resource_deploy_key.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					pattern := "ssh-rsa AAAA[0-9A-Za-z+/]+[=]{0,3}( [^@]+@[^@]+)?"
```
Comment (cwood) [inline bitbucket/resource_deploy_key.go:51]: We support more than just rsa. So this will break if you try to do a ecdsa key.

**Comment by cwood** [comment]
My only real concern is that the validation string wont really work since we support more than just RSA.

**Comment by cwood** [comment]
Sorry to have negative information. But essentially we are gonna be removing all of the 1.0 API's soon(can't give specifics - watch for a blog post). Iv talked to the devs who maintain and work on the apis for us and they havnt come up with any specifics. But it is similar to users/groups.

**Comment by sjauld** [comment]
Sorry, are you saying that Atlassian is removing support for adding deploy keys via the API? I'm a little confused...

**Comment by pdecat** [comment]
On the same stance here, confused, as the same arguments were invoked to decline #6.

**Comment by cwood** [comment]
Essentially bitbucket will not be implementing either of these apis for the
2.0 version. And I talked to our team who maintains these apis. And they
are going to deprecate these apis. So I don’t want to add these features
then have to remove them later with no real plan to migrate.

**Comment by sjauld** [comment]
So to confirm, Atlassian will remove the existing feature enabling the
programmatic management of deploy keys?

**Comment by cwood** [comment]
Yea unfortunately. I am not happy about this either. Your best bet would to make a site/master issue and ask for a 2.0 version of the deployment keys. I asked our API team again and they confirmed it.

**Comment by erikvanzijst** [comment]
> So to confirm, Atlassian will remove the existing feature enabling the programmatic management of deploy keys?

This is correct.

We don't yet have a date for when we turn off these APIs, but we will not be creating copies of these in 2.0. I've left a more complete motivation on the [site/master ticket](https://bitbucket.org/site/master/issues/15196/removal-of-10-api#comment-41240017).

**Comment by cwood** [comment]
Thanks for clarifying Erik

**Pull request closed without merging**